brainiak/eventseg/event.py

```diff
@@ -0,0 +1,356 @@
+        """Compute the log probability of each observed timepoint being
+        generated by the Gaussian distribution for each event pattern
+
+        Parameters
+        ----------
+        data: voxel by time ndarray
+            fMRI data on which to compute log probabilities
+
+        mean_pat: voxel by event ndarray
+            Centers of the Gaussians for each event
+
+        var: float or 1D array of length equal to the number of events
+            Variance of the event Gaussians. If scalar, all events are
+            assumed to have the same variance
+
+        Returns
+        -------
+        logprob : time by event ndarray
+            Log probability of each timepoint under each event Gaussian
+        """
+
+        n_vox = data.shape[0]
+        t = data.shape[1]
+
+        # z-score both data and mean patterns in space, so that Gaussians
+        # are measuring Pearson correlations and are insensitive to overall
+        # activity changes
+        data = stats.zscore(data, axis=0, ddof=1)
+        mean_pat = stats.zscore(mean_pat, axis=0, ddof=1)
+
+        logprob = np.empty((t, self.n_events))
+
+        if type(var) is not np.ndarray:
+            var = var * np.ones(self.n_events)
+
+        for k in range(self.n_events):
+            logprob[:, k] = -0.5 * n_vox * np.log(
+                2 * np.pi * var[k]) - 0.5 * np.sum(
+                (data.T - mean_pat[:, k]).T ** 2, axis=0) / var[k]
+
+        logprob /= n_vox
+        return logprob
+
+    def _forward_backward(self, logprob):
+        """Given the log probability of each timepoint being generated by
+        each event, run the HMM forward-backward algorithm to find the
+        probability that each timepoint belongs to each event (based on the
+        transition priors in p_start, p_end, and P)
+
+        Parameters
+        ----------
+        logprob : time by event ndarray
+            Log probability of each timepoint under each event Gaussian
+
+        Returns
+        -------
+        log_gamma : time by event ndarray
+            Log probability of each timepoint belonging to each event
+
+        ll : float
+            Log-likelihood of fit
+        """
+        t = logprob.shape[0]
+        logprob = np.hstack((logprob, float("-inf") * np.ones((t, 1))))
+
+        # Initialize variables
+        log_scale = np.zeros(t)
+        log_alpha = np.zeros((t, self.n_events + 1))
+        log_beta = np.zeros((t, self.n_events + 1))
+
+        # Forward pass
+        for t in range(t):
+            if t == 0:
+                log_alpha[0, :] = self._log(self.p_start) + logprob[0, :]
+            else:
+                log_alpha[t, :] = self._log(np.exp(log_alpha[t - 1, :])
+                                            .dot(self.P)) + logprob[t, :]
+
+            log_scale[t] = np.logaddexp.reduce(log_alpha[t, :])
+            log_alpha[t] -= log_scale[t]
+
+        # Backward pass
+        log_beta[-1, :] = self._log(self.p_end) - log_scale[-1]
```

Comment:
log_scale was calculated based on log_alpha in line 255, but not recalculated here based on log_beta, but it . I wonder if you need to do something similar as in line 255-256 but for log_beta. But since line 270 directly uses the log_scale calculated in line 255, perhaps you actually did not want to include anything from the backward pass in the ll? Sorry I do not know much of the forward-backward algorithm.